examples/experimental/dagster-airlift/dagster_airlift/core/airflow_cacheable_assets_def.py

```diff
@@ -0,0 +1,377 @@
+                # the leaf assets in each dag.
+                if (
+                    migration_state.get_migration_state_for_task(
+                        task_info.dag_id, task_info.task_id
+                    )
+                    is True
+                ):
+                    # Are there cases in airflow for which we'd want to relax this constraint?
+                    check.invariant(
+                        isinstance(correctly_typed_asset, AssetsDefinition),
+                        "If an asset is migrated, its compute function must exist in dagster.",
+                    )
+                    assets_def = cast(AssetsDefinition, correctly_typed_asset)
+                    check.invariant(
+                        assets_def.is_executable, "If an asset is migrated, it must be executable."
+                    )
+                    # For all specs in the assets def, we will create new specs with task-level metadata.
+                    # We will also build up a dictionary representing the global asset dependency, as well as all the asset keys
+                    # present in each dag.
+                    for spec in assets_def.specs:
+                        migrated_status_per_asset_key[spec.key.to_user_string()] = True
+                        all_asset_keys_per_dag_id[task_info.dag_id].append(
+                            spec.key.to_user_string()
+                        )
+                        for dep in spec.deps:
+                            downstreams_asset_dependency_graph[
+                                dep.asset_key.to_user_string()
+                            ].append(spec.key.to_user_string())
+                            upstreams_asset_dependency_graph[spec.key.to_user_string()].append(
+                                dep.asset_key.to_user_string()
+                            )
+                        keys_by_dag_and_task_id[task_info.dag_id][task_info.task_id].append(
+                            spec.key.to_user_string()
+                        )
+                        task_info_by_asset_key[spec.key.to_user_string()] = dict(task_info)
+                        # Construct the cacheable spec with migration information.
+                        new_spec = CacheableAssetSpec(
+                            asset_key=spec.key.to_user_string(),
+                            description=spec.description,
+                            metadata={
+                                **spec.metadata,
+                                **task_level_metadata,
+                                "Migrated": True,
+                                "Instigated by Task ID": task_info.task_id,
+                            },
+                            tags={**spec.tags},
+                            deps=[dep.asset_key.to_user_string() for dep in spec.deps],
+                            group_name=spec.group_name,
```

Comment:
it would be very very nice to break up this function. the "dual state" of the `asset` variable is very strange. as well.

My recommendation:

Break this into two sub-functions, one for the migrated case, the other in the unmigrated case. That way you will have fresh namespaces of variable names and will not need things like "`correctly_typed_asset`"

You'll probably need a tuple that represents all the data you compute on a per asset basis.

```python
class AssetData(NamedTuple):
    key: AssetKey
    upstream_assets: List[AssetKey]
    downstream_asset: List[AssetKey]
    task_info: TaskInfo
    specs: List[CacheableAssetSpec]
    migrated_status: bool
```

And then you'll need to stash that in the "extra metadata" section.